kube-runtime/src/coordination.rs

```diff
@@ -0,0 +1,603 @@
+                );
+                self.update_state(None);
+            }
+        };
+    }
+
+    /// Fetch the target lease from the API, and update observation info as needed.
+    #[tracing::instrument(level = "debug", skip_all)]
+    async fn k8s_fetch_lease_and_update(&mut self) -> Result<()> {
+        // Attempt to fetch the target lease, updating our last observed info on the lease.
+        let lease_opt = timeout(self.config.api_timeout, self.api.get_opt(&self.config.name))
+            .await
+            .map_err(|_err| Error::TimeoutError)?
+            .map_err(Error::ClientError)?;
+
+        // Lease info returned from the API, update observation info.
+        if let Some(lease) = lease_opt {
+            self.update_state(Some(lease));
+            return Ok(());
+        }
+        Ok(())
+    }
+
+    /// Attempt to acquire or renew the target lease.
+    #[tracing::instrument(level = "debug", skip_all, err)]
+    #[allow(clippy::cast_possible_truncation)]
+    async fn try_acquire_or_renew(&mut self) -> Result<()> {
+        tracing::debug!("try_acquire_or_renew");
+        // 1. Fetch the current state of the lease if following or standby.
+        if matches!(&self.state, State::Following { .. } | State::Standby) {
+            self.k8s_fetch_lease_and_update().await?;
+        }
+
+        // 2. Determine what type of update needs to be made to the lease.
+        // If following, and the lease is not expired (according to our own local time records), then no-op.
+        if matches!(&self.state, State::Following { .. }) && !self.is_lease_expired() {
+            return Ok(());
+        }
+        // Else, we need to patch the lease. Build up changeset.
+        let now = chrono::Utc::now();
+        let mut lease = self.state.get_lease().cloned().unwrap_or_default();
+        lease
+            .meta_mut()
+            .name
+            .get_or_insert_with(|| self.config.name.clone());
+        lease
+            .meta_mut()
+            .namespace
+            .get_or_insert_with(|| self.config.namespace.clone());
+        let spec = lease.spec.get_or_insert_with(Default::default);
+        spec.lease_duration_seconds = Some(self.config.lease_duration.as_secs_f32() as i32);
+        spec.renew_time = Some(MicroTime(chrono::Utc::now()));
+        if matches!(&self.state, State::Following { .. } | State::Standby) {
+            spec.holder_identity = Some(self.config.identity.clone());
+            spec.acquire_time = Some(MicroTime(now));
+            spec.lease_transitions = Some(spec.lease_transitions.map_or(0, |val| val + 1));
+        }
+        lease.metadata.managed_fields = None; // Can not pass this along for update.
+
+        // 3. Now we need to create or patch the lease in K8s with the updated lease value here.
+        let lease_res = if let State::Standby = &self.state {
+            timeout(
+                self.config.api_timeout,
+                self.api.create(&Default::default(), &lease),
```

Comment:
If we are using server-side apply below, we should use it here as well. No need to use `.create`.

Reply:
Yeah, if SSA is used for a given object it should typically be the _only_ (non-delete) write-op your controller uses for it.

Reply:
Agreed, will do. So far, this works as expected because this is only ever used in a pristine cluster / deployment where the lease does not yet exist. All subsequent updates are SSA w/ force=true, and give the expected result regardless.

Reply:
Tested and done.